### ISSUE TITLE
Allow user to set maximum weight value against which all other weights are normalized

### ISSUE THREAD

**Issue by kdahlquist** [dondi/GRNsight#264]
In the GRNmap meeting today, @GraceJohnson made a feature request regarding the display of weighted edges.

Right now, since each graph is normalized to the max weight of its own adjacency matrix, it is not really fair to compare two graphs with each other because they were normalized differently.

If we allowed the user to set the max value for normalization, then the user could compare two different graphs on the same scale.

This is a variant of the idea to somehow have a non-normalized scale for representing edge thicknesses as an alternate to the normalized scale.  This might be easier to implement.

**Comment by kdahlquist** [comment]
Just another idea--maybe we might want an option to transform the weights on a log scale which might help with some types of visualizations.

**Comment by kdahlquist** [comment]
This is actually a duplicate with an issue that @GraceJohnson created last summer, #218.  I'm going to close the other one and keep this one because it has a little more detail in its description.

**Comment by kdahlquist** [comment]
Another idea related to this:  allow the user to choose a threshold at which point the edges get greyed out.  Currently, the program automatically does this at 5% on either side of zero, but we could allow the users to choose another threshold (like 2.5%, 10%, etc.), possibly through a slider.

**Comment by kdahlquist** [comment]
See an example of such a slider at giant-api-biojs:  http://www.biojs.io/d/giant-api-biojs
![giant-api-biojs](https://cloud.githubusercontent.com/assets/6463867/17193606/c43af224-5408-11e6-9014-5d73eaa45c82.png)

**Comment by kdahlquist** [comment]
@anuvarsh is going to first work on the option display in the UI.  First we want to display the value being used, perhaps as an editable field, so the user can type a number there.

**Comment by dondi** [comment]
Upon inspection of the code, the variable to modify is `totalScale`, and here the `domain` of `totalScale` is what will be unhardcoded from its current value (which is the extent of edge weights in the GRN).

**Comment by anuvarsh** [comment]
Reading [this article](https://www.dashingd3js.com/d3js-scales) to get more information about scaling in D3. 

Leaving the link here so I can quickly find the article in the future.

**Comment by anuvarsh** [comment]
I figured out how to update the normalization of the weights using d3. If the user provides a min and max value for their data, coding in the corresponding normalization should be fairly straightforward. 

Currently, a very basic front end with two input boxes has been pushed onto beta. Next, I am going to try to present the current min and max data points as placeholders within the text. Taking a break for now so I can help @mihirsamdarshi with the poster for Saturday.

**Comment by kdahlquist** [comment]
So does this mean that we are changing the way that the normalization is being done?  In my mind, it seems that we only need to know the max, not the min.  I.e., every weight is divided by the max (absolute value) weight.  0 weights are still zero, and every other one is normalized to that max weight.  Why do we need to know the min?  Isn't it just 0 anyway?

**Comment by kdahlquist** [comment]
There is an edge case of a fully connected graph where the min would not be 0, so we need the min.  I suggest that the field be there, prefilled with 0 and a note to the user that says "unless you have a fully connected graph, this value should be 0."

@NAnguiano points out that need to disallow a max lower than a min.

**Comment by kdahlquist** [comment]
Just another note from the meeting; I was under the impression that we wrote our own normalization routine, but D3 is handling it along with binning the weights into edge thicknesses.

**Comment by dondi** [comment]
Due to the unexpected difficulties with #285, this did not see movement this week.

**Comment by dondi** [comment]
@anuvarsh just started but at the meeting we reviewed the portions of the _graph.js_ code that would likely impact this new feature.

**Comment by kdahlquist** [comment]
@anuvarsh is still working on this.

* code can detect the normalization factors from uploaded graph
* can take a user-provided normalization factors
* But, the graph is not refreshing
    * she will write some text in the UI that warns the user that the graph will reload if they change the normalization factors

**Comment by dondi** [comment]
@anuvarsh expects to have this finished by tonight. @dondi discussed the issue of minimum/maximum vs. single normalization with her and it was concluded that a standardized minimum of zero should suffice and will in fact be easier to code. @dondi can discuss the potential difference in visualization with @kdahlquist but the impact is expected to be minimal.

**Comment by anuvarsh** [comment]
Normalization is now responding to user input data. There is some weird functionality where if the user tries to normalize the data to a number smaller than the maximum value in the network, d3 no longer scales the data to the range.

![image](https://cloud.githubusercontent.com/assets/8509674/25466911/2c864ba6-2ac0-11e7-838f-38c4ccc9a279.png)

Initial array (adjustedWeights): [0.005287, 0.00263, 0.1447, 0.1233, 0.4212, 0.04996, 0.01562, 0.3795, 0.0000582, 0.0168, 0.5054, 0.006903, 0.04832, 0.3817, 0.4142, 0.1078, 0.1569, 0.1993, 0.04304, 0.5417, 0.09537, 0.1979, 0.004389, 0.06778, 1.001, 0.002523, 0.1261, 0.3293, 0.07523, 0.0497, 0.06664, 0.09177, 0.00153, 0.004437, 0.0977, 1.862, 0.0002295, 0.1299, 0.09312, 0.0459, 0.03359, 0.008698, 0.001634, 0.01592, 0.202, 0.001244, 0.2682, 0.001939, 0.1027, 0.1528, 0]

User provided max value for normalization: 0.5

Returned array (newScaledData): [0, 0.5, -26.23503951825367, -22.20794128716598, -78.26740684983065, -8.406661648475724, -1.9444862627022959, -70.42021076401957, 0.9839668799397818, -2.1665412118931124, -94.1123447497177, -0.3041023710952202, -8.098042905532557, -70.83421151674821, -76.95013172751224, -19.291117802032367, -28.530861874294317, -36.50978547233722, -7.104441098983817, -100.9433571697403, -16.952013549115545, -36.24633044787354, 0.1689875799774182, -11.760067745577722, -187.37542340986073, 0.5201354911554384, -22.734851336093335, -60.9734663153933, -13.162024840045165, -8.357734286789613, -11.545540082800152, -16.274557771923224, 0.7070003763643207, 0.15995483628152052, -17.39047798268724, -349.40026345502446, 0.9517312758750471, -23.4499435453519, -16.528603688370342, -7.64264207753105, -5.326119683853971, -0.6418893488897252, 0.6874294316898758, -2.0009409108016563, -37.01787730523147, 0.760820474219044, -49.475536319156944, 0.6300338727888597, -18.331388784343243, -27.759315016936394, 0.9949190816710576]

The returned array should have values only between -0.5 to 0.5, so I'm not sure why I'm not seeing this. If anyone sees anything, please let me know!

**Comment by kdahlquist** [comment]
I don't have any insight into @anuvarsh's last comment, but I just wanted to say that I had thought since the beginning that the minimum was 0, so was confused why it was anything but that.  The documentation I wrote for the feature assumes this, actually.  So this change is actually encoding the original spec.

It just goes to show that there are always hidden assumptions to tease out.  Looking forward to seeing the new version.

**Comment by NAnguiano** [comment]
@anuvarsh I quickly looked at this and I think I found the issue. 

The normalized weights aren't actually being used anywhere. The original weights are used to set the weights of the links so they're correct on mouseover. But because of that, the original weights are what are used to draw the graph. So as a result, even though you're limiting the values to be in a range from 0 to 0.5, it's still trying to draw it as though it's from -1.6 to 1.

If you change the weights to use the normalized weights, the graph seems to draw within the range properly.

![screen shot 2017-04-27 at 12 50 27 pm](https://cloud.githubusercontent.com/assets/6444156/25501585/203353de-2b48-11e7-98c8-3d1733b85709.png)

As you can see, there's something weird going on with the bars, but other than that things are looking more within range. The weights on mouseover are also wrong now. So it's not a perfect solution, but at least the reason I think was found.

This is what I did: 
![screen shot 2017-04-27 at 12 51 33 pm](https://cloud.githubusercontent.com/assets/6444156/25501633/4bd38086-2b48-11e7-8b56-064f443b33fd.png)

I'm not sure if it's relevant or not, but I think that the domains are also supposed to be using d3.extent(array) rather than just the array. The domain is two values, so passing in the entire array just leads to the first two values of the array being used. It didn't make a difference here, but you may want to do that instead. You can see I changed it in the code above.

**Comment by kdahlquist** [comment]
I just wanted to make a note that the file you should use for testing should be an artificially constructed file with weight values that chosen specifically for testing purposes.

**Comment by anuvarsh** [comment]
@NAnguiano, if possible, could you push that to my branch? I tried to add that to my code according to your picture and description but about 10 new issues came up. 

Updates for @kdahlquist and @dondi, yesterday night when I was working on this last little bit, I reached a hurdle that made me realize that I had to go back and basically completely reconstruct how I was going to go about this problem which ended up revealing some d3 problems. Prior to yesterday's work session, all of the data was properly normalized to the user provided data, but I could not update the graph. Now that the graph is uploading properly, I am realizing that there are some interim variables that need to be completely changed for normalization. Updated goal is to have this done by Sunday night.

**Comment by NAnguiano** [comment]
@anuvarsh Sure! I pushed it. I also quickly found and fixed the issue with the repressor bars. This is what the 0.5 normalized graph looks like now.
![screen shot 2017-04-27 at 5 22 00 pm](https://cloud.githubusercontent.com/assets/6444156/25509466/126465f6-2b6e-11e7-9c51-cfd572bdad59.png)

**Comment by anuvarsh** [comment]
Wow! Yay! Thank you! Now working on making sure extreme values are handled.

**Comment by anuvarsh** [comment]
@kdahlquist, should we restrict the values a user can normalize data with to be within some parameters? If so, any recommendations on what those parameters should be?

I thought that we should allow all numbers greater than 0, but I can see how there might be an argument to set a minimum value that is slightly higher than 0, and also potentially setting a maximum value.

**Comment by kdahlquist** [comment]
I thought we didn't need the minimum value 
anymore, since it is always going to be zero.

We probably do need a maximum for the maximum 
value.  I'm going to arbitrarily suggest 
1000.  That should be sufficiently large without 
being insane.  Please make this easily changeable 
in the code so if we need to change it later, it's not a big deal.

At 08:17 PM 4/27/2017, you wrote:

**Comment by kdahlquist** [comment]
This is how I think that the normalization is/should be working.

1. GRNsight finds all the nonzero weights.
2. It takes the absolute value (to get rid of the negative signs).
3. It then finds the maximum value
4. It divides every other non-zero weight by this maximum, returning the normalized weight values, which are now all going to be >0 and <=1 for the positive weights or <0 and >=(-1) for the negative weights.
5. Since there are 12 bins for edge thicknessess, here is what I think the bins should range from for normalized weights (to 4 decimal places, which is what we are rounding to in the display since 12 doesn't go into 1 very nicely) from thinnest to thickest (this is just considering magnitude):
* 0 - 0.0833
* 0.0834 - 0.1666
* 0.1667 - 0.2499
* 0.2500 - 0.3332
* 0.3333 - 0.4165
* 0.4166 - 0.4998
* 0.4999 - 0.5831
* 0.5832 - 0.6664
* 0.6665 - 0.7497
* 0.7498 - 0.8330
* 0.8331 - 0.9163
* 0.9164 - 1.0000

So, in this scheme, I just don't see why we need to even know the minimum nonzero weight value; it is not part of any calculation.  All normalized weights between 0 and 0.0500 are grey and the thinnest bin.  Normalized weights between 0.0501 and 0.0833 are their respective color (magenta or cyan) and the thinnest bin.

I had assumed all along that this the way that it works, but if it isn't then I need to understand what the actual mechanism is.

**Comment by dondi** [comment]
I agree with this, and the normalization value essentially customizes that maximum. @anuvarsh please describe a scenario where a custom minimum needs to be selected so we can understand that need better. After the Wednesday meeting I thought that had been settled. Thanks!

**Comment by anuvarsh** [comment]
I'm sorry, I did a poor job clarifying. 

We are only taking in "maximum" normalization values from the user. This "maximum" value acts as the upper value of the range we want the matrix data to fit into. My question was in regards to the smallest "maximum" value (or normalization favor/ range) we would allow. Is the user allowed to input something like 10^-12 as a normalization factor?

Currently I have the code set such that *any* number greater than 0 is allowed to be the normalization factor/ range. Would it make sense to say the smallest value we will allow is anything besides 0?

**Comment by dondi** [comment]
Ah OK I understand now. So to paraphrase, we are looking for the acceptable range of the normalization value. @kdahlquist has chimed in 1000 on the high end.

With that in mind, I don’t see any reason not to allow a “minimum maximum” of zero. This effectively turns off thickness visualization on the GRN because all edge values will land in the same “bin.” This might actually be an interesting mode, because with a normalization value of zero, all edges will be the same thickness and the only differentiator will be the color.

**Comment by anuvarsh** [comment]
Sounds good to me! Working on that now.

**Comment by kdahlquist** [comment]
I agree with @dondi.  Sorry for the previous misunderstanding.

**Comment by anuvarsh** [comment]
I made a couple CSS and Jade changes to my code before pulling the changes that @NAnguiano made above. After pulling those changes, however, I'm getting errors on every file that I am trying to upload including demos. The errors are just bad request errors for the demos and missing network sheet errors for every other file. 

I am going to try to fuss around and see if I can figure out what is going on with the code that was pushed, but unless I figure out something soon I'm pretty sure this means I have to back track to my commits from last week and try to reincorporate Nicole's changes manually one by one. Kind of a bummer, but still trying to get this done by Wednesday.

**Comment by NAnguiano** [comment]
It's worth noting that @yshin4 had the same error that @anuvarsh is reporting, as noted on issue #455. I haven't had any problems so far, but I'm a bit concerned that this bug is now affecting two people... @yshin4 did you manage to find a reason or fix when it was happening to you? Does it persist across browsers, and with all extensions off?

**Comment by anuvarsh** [comment]
Tried to pull the updates from @NAnguiano, and run GRNsight locally and it still isn't working. For reference, this is the issue that I'm getting when I try to upload anything (this is when I tried to upload one of the demo files). 

![image](https://cloud.githubusercontent.com/assets/8509674/25648153/b805e8aa-2f7d-11e7-87f0-3061307dde7c.png)

I'm at a loss, I have no idea what is causing the problem. Perhaps me and @NAnguiano can meet up and figure it out?

**Comment by NAnguiano** [comment]
That is so weird. I just re-cloned your branch, instantiated everything back from 0, npm installed, and..
![screen shot 2017-05-02 at 9 31 30 pm](https://cloud.githubusercontent.com/assets/6444156/25648247/cc074474-2f7e-11e7-9cc1-99b8d5f0f95e.png)
it's working perfectly. We should definitely meet up if we can't figure this out. I honestly don't know why it wouldn't be working.

**Comment by dondi** [comment]
@NAnguiano I’ve posted in #455 for @anuvarsh and/or @yshin4 to find me if this is still happening to them. Happily this is not happening to you so you can focus on your stuff :)

**Comment by kdahlquist** [comment]
@anuvarsh has moved all of her in progress work to a new normalization branch since there are still some outstanding bugs that will prevent this feature from being in the new release.

We should hide the "Normalization factor" options for now.

**Comment by kdahlquist** [comment]
So, I'm looking at v2.0.17 and it's great to see some functionality I can play around with.

However, I don't think this implementation is ready for prime time and will still need to be hidden for the release.  Things I have observed:

* The thickness of the edges are behaving as expected:
    * I.e., if I set the normalization factor to something much larger than the max weight of the graph, all the lines get skinny.
    * If I set the normalization factor to something much smaller than the smallest weight of the graph, all the lines get fat.
* However, the color/grey part is not working as I had intended (although I probably never spelled this out explicitly).
    * The rules for when an edge is grey, should also be affected by the new normalization factor.  
    * Edges should always follow the rule that an edge should be grey if it's (normalized) weight is within 5% of 0.
    * That means that if I set the normalization factor to something much larger than the max weight of the graph, then all of the lines should get skinny *and* grey because now all the normalized weight values are withing 5% of 0.
    * That also means that if I set the normalization factor to something much smaller than the smallest weight on the graph, then all the lines should get fat and *none* of them should be grey because now none of the normalized weight values are within 5% of zero.
* Also, it looks like we are also going to need a "reset" button for the weights that will unapply the new normalization factor.
    * Right now, it's actually remembered if I reload the page, blowing away the current graph, which is great if you want to view several graphs in a row with the same normalization factor, but not so good if you don't want to do that.

Based on conversations we started in @dondi's office with @anuvarsh and @kdahlquist yesterday, there are a lot of design decisions that actually need to be made for the UI as well.

Progress has been made, though and that's good news!

**Comment by dondi** [comment]
OK, sounds good—the focus so far has been solely the first bullet so it’s good to see that this one is up and running now. I agree that the next steps include working on (and deciding on) the other two bullets.

This demo is running directly off the `normalization` branch so it is easy to switch away from it because it hasn't actually landed in the `beta` branch yet. Future work can continue on this branch (or someone’s fork of this branch).

**Comment by kdahlquist** [comment]
@mihirsamdarshi is going to take on this task.  Yesterday he started reviewing the code.  We need to have a group discussion with @dondi about next steps in the implementation and UI decisions.

**Comment by dondi** [comment]
I can do a Skype/Hangout call this week—either of those will be ideal because we would be able to screenshare and have discussions over code. For such a call I have a fair amount of flexibility, outside of commuting or other meetings; let me know when everyone’s available.

**Comment by mihirsamdarshi** [comment]
Running into a few problems/questions regarding implementation of the gray edges threshold. I was looking at implementing a slider for normalization, however two problems that I see is that I would potentially have to reload the graph for every input into the slider (meaning if you wanted to change the slider from 0.05 to 0.25, the graph would have to be reloaded 25 times). Second, the entire sliders.js file is purpose built to serve the Force Graph view options, although that isn't as big of a deal. I have two potential solutions to this which is just going with a text box, or still creating a slider, but creating a button to apply the gray threshold changes to the graph.

**Comment by dondi** [comment]
The first issue has a straightforward resolution: don’t “listen” for changes to the slider as it changes, but only when it “settles down” (i.e., the user lets it go). I’m not familiar with the specific slider widget currently in use, but in other frameworks the events are `input` for when you want to respond to changes as they occur, and `change` when you want to respond only after a value is resolved. Just go with the `change` event. No need for an Apply button.

Now the second issue is whether you can even reuse the existing sliders. If they are too closely bound to the force graph parameters (they shouldn’t be, but then again these were some of the earliest things that @NAnguiano did so she may not have been as aware of separation of concerns back then), just use a different slider. The latest HTML5 `input` element has a slider option, I believe, when `type="number"`. Look that up and consider using that. For that one, I know that the `change` event will do what you want.

The thing about the graph reloading is an on-going issue, kind of an artifact of how the D3 force graph is implemented. That may need to be a separate task in the future, for when we have greater in-depth familiarity with how the overall visualization code works.

**Comment by dondi** [comment]
@mihirsamdarshi This might get you started on the HTML5 `input` element as a slider:

    http://thenewcode.com/757/Playing-With-The-HTML5-range-Slider-Input

**Comment by kdahlquist** [comment]
@mihirsamdarshi reported some issues with not getting the Jade file to produce the desired HTML. @dondi noted that spacing is significant in Jade (now Pug) and this might be what's throwing him off. @mihirsamdarshi will take another stab at the edit with spacing in mind, then if it still doesn't cooperate, he will send the file to @dondi.

**Comment by mihirsamdarshi** [comment]
Got the spacing working, realized that the values in between the name of my div or my slider and the parentheses actually mattered, now just doing some changes to the javascript to help it run right.

**Comment by mihirsamdarshi** [comment]
Done with the technical aspects, now just working on formatting.

**Comment by mihirsamdarshi** [comment]
@dondi Could you take a look at line 274 in upload.jade on the normalization branch in my fork, I'm not sure how I can follow the previous formatting and make the output fall on the same line as the title, as is the format on the other sliders. Other than that formatting change, the feature is completed.

**Comment by dondi** [comment]
@mihirsamdarshi That type of formatting is not done in HTML, but in CSS. Most likely, the `output` element is defaulting to a `display: block` style. If you make its CSS be `display: inline` or `display: inline-block` then that should stay on the same line (as long as title is similarly styled).

I’m not sure how much CSS you know so far (and there are wrong and right ways of doing it), but in substance that would be the thing to do.

**Comment by eileenchoe** [comment]
Another question: Should the "Set Normalization Factor" and "Set Grey Line Threshold" disappear when an unweighted graph is loaded, since these options are only relevant for weighted graphs? This is the behavior that we currently have for "Hide/Show Edge Weights"

**Comment by kdahlquist** [comment]
My feeling on this one is that there is an argument to be made for having these visible so that the user can set them before s/he loads a new graph.  Let's have it be this way for now and then we can see if we like it when we play around with it some.

**Comment by eileenchoe** [comment]
Live on beta v2.1.4

**Comment by kdahlquist** [comment]
I like what I see so far.  I want to do a little more testing on some specific files that have a range of edge sizes before I am satisfied that all is behaving as expected.

In the meantime, let's think about some UI tweaks.

I think our practice of showing the allowable range of values will be helpful here as well.  Again, like with zoom, it doesn't make sense to have "0" as the minimum because that invokes a divide by zero error.  Since we basically show four decimals for weight values, the minimum should be 0.0001.  I think I told @anuvarsh at one point that the max should be 1000, so our range would be 0.0001 - 1000.  So the label could look something like:

Edge weight normalization factor (0.0001 - 1000): [box]

obviously, this is too long to fit on one line, so maybe it will fit on two.  

Then the box only needs to be as wide as it would take to type in 5 digits, we don't need the whole width that we have now.

I think we can just get rid of the text that says "*assumes a min. value of 0" which was a holdover from the old way we were thinking about this.

The "Set" and "Reset" buttons could go below then on the same horizontal line.

In terms of the Gray Threshold, I don't think we need the "Set Gray Threshold" header because it is redundant with the label above the slider and the slider does convey the idea that you can change the value.

In some comment, @eileenchoe asked whether these values should be between 0 and 1 or 0 and 100%.  I asked @nikavafadari and Nick and they though the 100% scale was better.  I agree with them and want to request this change.  So the label above the slider would say:

Gray Edge Threshold (0 - 100%): value

Note that I inserted the word "Edge" there.

I like that you added the tooltip to explain the slider (note that we need to add tooltips for some of the other options, separate issue forthcoming).

I suggest a slightly different wording to the tooltip.

"Increases/decreases the threshold at which edges appear gray. Edges with normalized weight values below this percentage are shown as gray."

I opened another issue #505 for us to comment on the spelling of gray/grey.

That's all for now.  Again, very excited to see this feature and play around with it more.

**Comment by kdahlquist** [comment]
@mihirsamdarshi, there are some changes you can make that are listed in the comment above, while I am reviewing the rest of the functionality.

**Comment by kdahlquist** [comment]
I've created some test files to look at the bins for edge thicknesses so that I can test this functionality more fully.  I'm still in the midst of looking at the thicknesses without actually changing the normalization factor, so I can't really comment on that yet.  I want to thoroughly test that first before messing with changing the normalization.  However, I've seen enough to say that the gray threshold is working great.

Again, @mihirsamdarshi, please proceed with the UI changes requested above while I complete the functional testing.  I may not actually be able to finish it all before we do a release, but I think this is working well enough that it can be released.

**Comment by mihirsamdarshi** [comment]
I fixed all the UI changes that were requested, the conversion to percentage is slightly more complicated, because of the way the code is written. Because we are trying to get the last pull request before the release done today, I think that my efforts are best directed to other issues right now.

**Comment by kdahlquist** [comment]
Declaring victory on this one and closing!  Any subsequent changes/bugs will get their own follow-up issues, like #518, changing gray threshold scale from 1 - 100% instead of 0 - 1.
